lib/langchain/tool_definition.rb

```diff
@@ -0,0 +1,209 @@
+# frozen_string_literal: true
+
+require "json"
+
+#
+# Extends a class to be used as a tool in the assistant.
+# A tool is a collection of actions (methods) used to perform specific tasks.
+#
+# == Usage
+#
+# 1. Extend your class with {Langchain::ToolDefinition}
+# 2. Use {#define_action} to define each action of the tool
+#
+# == Key Concepts
+#
+# - {#define_action}: Defines a new action (method) for the tool
+# - {ParameterBuilder#property}: Defines properties for the action parameters
+# - {ParameterBuilder#item}: Alias for {ParameterBuilder#property}, used for array items
+#
+# These methods support various data types and nested structures, allowing for flexible and expressive tool definitions.
+#
+# @example Defining a tool with various property types and configurations
+#   define_action :sample_action, description: "Demonstrates various property types and configurations" do
+#     property :string_prop, type: "string", description: "A simple string property"
+#     property :number_prop, type: "number", description: "A number property"
+#     property :integer_prop, type: "integer", description: "An integer property"
+#     property :boolean_prop, type: "boolean", description: "A boolean property"
+#     property :enum_prop, type: "string", description: "An enum property", enum: ["option1", "option2", "option3"]
+#     property :required_prop, type: "string", description: "A required property", required: true
+#     property :array_prop, type: "array", description: "An array property" do
+#       item type: "string", description: "Array item"
+#     end
+#     property :object_prop, type: "object", description: "An object property" do
+#       property :nested_string, type: "string", description: "Nested string property"
+#       property :nested_number, type: "number", description: "Nested number property"
+#     end
+#   end
+#
+module Langchain::ToolDefinition
+  # Defines an action for the tool
+  #
+  # @param method_name [Symbol] Name of the method to define
+  # @param description [String] Description of the action
+  # @yield Block that defines the parameters for the action
+  def define_action(method_name, description:, &)
+    action_schemas.add_action(method_name:, description:, &)
+  end
+
+  # Returns the ActionSchemas instance for this tool
+  #
+  # @return [ActionSchemas] The ActionSchemas instance
+  def action_schemas
+    @action_schemas ||= ActionSchemas.new(tool_name)
+  end
+
+  # Returns the snake_case version of the class name as the tool's name
+  #
+  # @return [String] The snake_case version of the class name
+  def tool_name
+    @tool_name ||= name.gsub(/([A-Z])/, '_\1').gsub(/^_/, "").gsub("::", "").downcase
```

Comment:
There could be a bug here. Example:
```ruby
class Langchain::Tool::API1Interface; end

irb(main):023> "Langchain::Tool::API1Interface".gsub(/([A-Z])/, '_\1').gsub(/^_/, "").gsub("::", "").downcase
=> "langchain_tool_a_p_i1_interface"
```

Reply:
Thanks for pointing this out! It's not really a problem, the mechanism still works. But if we decide to go with the automatic naming approach, it's better to fix it for easier reading.

Reply:
The (possible) problem here is that the LLM uses this name as context as well, so if the name is malformed then it might mess up the LLM. Obviously the best function name is one that is descriptive to what the function actually does.

Reply:
Ok, I had missed this point! Thank you

Reply:
Fixed inspired by the `underscore` method from `Active Support Inflector`